lib/rubocop/cop/rails/duplicate_association.rb

```diff
@@ -0,0 +1,67 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Rails
+      # This cop looks for associations that have been defined multiple times in the same file.
+      #
+      # @example
+      #
+      #   # bad
+      #   belongs_to :foo
+      #   belongs_to :bar
+      #   has_one :foo
+      #
+      #   # good
+      #   belongs_to :foo
+      #   has_one :bar
+      #
+      class DuplicateAssociation < Base
+        include RangeHelp
+        extend AutoCorrector
+
+        MSG = "Association `%<name>s` is defined multiple times. Don't repeat associations in a model."
+
+        def_node_matcher :association, <<~PATTERN
+          (send nil? ${:belongs_to :has_one :has_many :has_and_belongs_to_many} ({sym str} $_) ...)
+        PATTERN
+
+        def on_class(class_node)
+          offenses(class_node).each do |name, nodes|
+            length = nodes.length
+
+            nodes.each.with_index do |node, i|
+              add_offense(node, message: format(MSG, name: name)) do |corrector|
+                corrector.remove(range_by_whole_lines(node.source_range, include_final_newline: true)) if i < length - 1
+              end
+            end
+          end
+        end
+
+        private
+
+        def offenses(class_node)
+          class_send_nodes(class_node).select { |node| association(node) }
+                                      .group_by { |assoc| association_name(assoc) }
```

Comment:
`association` node matcher captures (`$`) type and name of associations.
If you call `association(assoc)`, it will return something like `[:belongs_too, :foo]`.
You may simplify code a bit by:
 - removing capture (`$`) from `${:belongs_to`, as we don't use association type anymore
 - using `group_by` by the same `association` and remove `association_name` method

Reply:
![image](https://user-images.githubusercontent.com/6916/147555457-0eba77d4-063f-4a17-ab41-e6f6fb40aa2e.png)

Reply:
Makes sense. Thanks!